src/gtc/daceir.py

```diff
@@ -0,0 +1,892 @@
+    def mask_is_boolean_field_expr(cls, v: Expr) -> Expr:
+        if v.dtype != common.DataType.BOOL:
+            raise ValueError("Mask must be a boolean expression.")
+        return v
+
+
+class HorizontalRestriction(common.HorizontalRestriction[Stmt], Stmt):
+    pass
+
+
+class UnaryOp(common.UnaryOp[Expr], Expr):
+    pass
+
+
+class BinaryOp(common.BinaryOp[Expr], Expr):
+    _dtype_propagation = common.binary_op_dtype_propagation(strict=True)
+
+
+class TernaryOp(common.TernaryOp[Expr], Expr):
+    _dtype_propagation = common.ternary_op_dtype_propagation(strict=True)
+
+
+class Cast(common.Cast[Expr], Expr):  # type: ignore
+    pass
+
+
+class NativeFuncCall(common.NativeFuncCall[Expr], Expr):
+    _dtype_propagation = common.native_func_call_dtype_propagation(strict=True)
+
+
+class While(common.While[Stmt, Expr], Stmt):
+    pass
+
+
+class ScalarDecl(Decl):
+    pass
+
+
+class LocalScalar(ScalarDecl):
+    pass
+
+
+class SymbolDecl(ScalarDecl):
+    def to_dace_symbol(self):
+        return get_dace_symbol(self.name, self.dtype)
+
+
+class Temporary(FieldDecl):
+    pass
+
+
+class ComputationNode(LocNode):
+    # mapping connector names to tuple of field name and access info
+    read_memlets: List[Memlet]
+    write_memlets: List[Memlet]
+
+    @validator("read_memlets", "write_memlets")
+    def unique_connectors(cls, node: List[Memlet]):
+        conns: Dict[SymbolRef, SymbolRef] = dict()
+        for memlet in node:
+            conns.setdefault(memlet.field, set())
+            if memlet.connector in conns[memlet.field]:
+                raise ValueError(f"Found multiple Memlets for connector '{memlet.connector}'")
+            conns[memlet.field].add(memlet.connector)
+        return node
+
+    @property
+    def read_fields(self):
+        return set(ml.field for ml in self.read_memlets)
+
+    @property
+    def write_fields(self):
+        return set(ml.field for ml in self.write_memlets)
+
+    @property
+    def input_connectors(self):
+        return set(ml.connector for ml in self.read_memlets)
+
+    @property
+    def output_connectors(self):
+        return set(ml.connector for ml in self.write_memlets)
+
+
+class IterationNode(Node):
+    grid_subset: GridSubset
+
+
+class Tasklet(ComputationNode, IterationNode):
+    stmts: List[Union[LocalScalar, Stmt]]
+    grid_subset: GridSubset = GridSubset.single_gridpoint()
+
+
+class DomainMap(ComputationNode, IterationNode):
+    index_ranges: List[Range]
+    schedule: MapSchedule
+    computations: List[Union[Tasklet, "DomainMap", "NestedSDFG"]]
+
+
+class ComputationState(IterationNode):
+    computations: List[Union[Tasklet, DomainMap]]
+
+
+class DomainLoop(IterationNode, ComputationNode):
+    axis: Axis
+    index_range: Range
+    loop_states: List[Union[ComputationState, "DomainLoop"]]
+
+
+class NestedSDFG(ComputationNode):
+    label: SymbolRef
+    field_decls: Dict[SymbolRef, FieldDecl]
+    symbol_decls: Dict[SymbolRef, SymbolDecl]
```

Comment:
The other IRs use a style where these are lists, and there is a symbol table (`symtable_` attribute) is implicitly defined. The visitors that need decls add the visitor context by adding

```
contexts = (SymbolTableTrait.symtable_merger,)
```

to the class. Is there a reason to not use this system for the daceir?

Side note: soon the method for adding the visitor context will have to change across gt4py, but for now this is the consistent approach.